base/ctf/msctf/context.cpp

```diff
@@ -0,0 +1,1012 @@
+    return Range_Constructor(this, m_pITextStoreACP, cookie->lockType, 0, 0, ppStart);
+}
+
+STDMETHODIMP
+CContext::GetEnd(
+    _In_ TfEditCookie ec,
+    _Out_ ITfRange **ppEnd)
+{
+    EditCookie *cookie;
+    LONG end;
+
+    TRACE("(%p) %i %p\n", this, ec, ppEnd);
+
+    if (!ppEnd)
+        return E_INVALIDARG;
+
+    *ppEnd = NULL;
+
+    if (!m_connected)
+        return TF_E_DISCONNECTED;
+
+    if (get_Cookie_magic(ec) != COOKIE_MAGIC_EDITCOOKIE)
+        return TF_E_NOLOCK;
+
+    if (!m_pITextStoreACP)
+    {
+        FIXME("Context does not have a ITextStoreACP\n");
+        return E_NOTIMPL;
+    }
+
+    cookie = (EditCookie *)get_Cookie_data(ec);
+    m_pITextStoreACP->GetEndACP(&end);
+
+    return Range_Constructor(this, m_pITextStoreACP, cookie->lockType, end, end, ppEnd);
+}
+
+STDMETHODIMP CContext::GetActiveView(_Out_ ITfContextView **ppView)
+{
+    FIXME("STUB:(%p)\n", this);
+    return E_NOTIMPL;
+}
+
+STDMETHODIMP CContext::EnumViews(_Out_ IEnumTfContextViews **ppEnum)
+{
+    FIXME("STUB:(%p)\n", this);
+    return E_NOTIMPL;
+}
+
+STDMETHODIMP CContext::GetStatus(_Out_ TF_STATUS *pdcs)
+{
+    TRACE("(%p) %p\n", this, pdcs);
+
+    if (!m_connected)
+        return TF_E_DISCONNECTED;
+
+    if (!pdcs)
+        return E_INVALIDARG;
+
+    if (!m_pITextStoreACP)
+    {
+        FIXME("Context does not have a ITextStoreACP\n");
+        return E_NOTIMPL;
+    }
+
+    m_pITextStoreACP->GetStatus(&m_documentStatus);
+
+    *pdcs = m_documentStatus;
+
+    return S_OK;
+}
+
+STDMETHODIMP CContext::GetProperty(
+    _In_ REFGUID guidProp,
+    _Out_ ITfProperty **ppProp)
+{
+    FIXME("STUB:(%p)\n", this);
+    return E_NOTIMPL;
+}
+
+STDMETHODIMP CContext::GetAppProperty(
+    _In_ REFGUID guidProp,
+    _Out_ ITfReadOnlyProperty **ppProp)
+{
+    FIXME("STUB:(%p)\n", this);
+    return E_NOTIMPL;
+}
+
+STDMETHODIMP CContext::TrackProperties(
+    _In_ const GUID **prgProp,
+    _In_ ULONG cProp,
+    _In_ const GUID **prgAppProp,
+    _In_ ULONG cAppProp,
+    _Out_ ITfReadOnlyProperty **ppProperty)
+{
+    FIXME("STUB:(%p)\n", this);
+    return E_NOTIMPL;
+}
+
+STDMETHODIMP CContext::EnumProperties(_Out_ IEnumTfProperties **ppEnum)
+{
+    FIXME("STUB:(%p)\n", this);
+    return E_NOTIMPL;
+}
+
+STDMETHODIMP CContext::GetDocumentMgr(_Out_ ITfDocumentMgr **ppDm)
+{
+    TRACE("(%p) %p\n", this, ppDm);
+
+    if (!ppDm)
+        return E_INVALIDARG;
+
+    *ppDm = m_manager;
+    if (!m_manager)
+        return S_FALSE;
+
+    m_manager->AddRef();
+    return S_OK;
+}
+
+STDMETHODIMP CContext::CreateRangeBackup(
+    _In_ TfEditCookie ec,
+    _In_ ITfRange *pRange,
+    _Out_ ITfRangeBackup **ppBackup)
+{
+    FIXME("STUB:(%p)\n", this);
+    return E_NOTIMPL;
+}
+
+STDMETHODIMP CContext::AdviseSink(
+    _In_ REFIID riid,
+    _In_ IUnknown *punk,
+    _Out_ DWORD *pdwCookie)
+{
+    TRACE("(%p) %s %p %p\n", this, debugstr_guid(&riid), punk, pdwCookie);
+
+    if (cicIsNullPtr(&riid) || !punk || !pdwCookie)
```

Comment:
I bet the `cicIsNullPtr(&riid)` is 99.99% a hack, since our other C++ shell code dealing with REFIIDs don't have that much hack for this.

Reply:
`!&riid` generates warnings.
Skipping null check will generate test failures.